constraint-solver/src/solver/constraint_splitter.rs

```diff
@@ -0,0 +1,565 @@
+        let candidate = &components[index];
+        let rest = components
+            .iter()
+            .enumerate()
+            // Filter out the candidate itself and all zero components
+            // because we set split components to zero when we extract them.
+            .filter(|(i, component)| *i != index && !component.is_zero())
+            .map(|(_, comp)| (comp.clone() / candidate.coeff).normalize())
+            .collect_vec();
+        // The original constraint is equivalent to
+        // `candidate.expr + rest + constant / candidate.coeff = 0`.
+        // Try to find the unique value for `candidate.expr` in this equation.
+        if let Some(solution) = find_solution(
+            &candidate.expr,
+            rest,
+            constant / candidate.coeff,
+            range_constraints,
+        ) {
+            // We now know that `candidate.expr = solution`, so we add it to the extracted parts.
+            extracted_parts.push(AlgebraicConstraint::assert_eq(
+                candidate.expr.clone(),
+                GroupedExpression::from_number(solution),
+            ));
+            // We remove the candidate (`candidate.coeff * candidate.expr`) from the expression.
+            // To balance this out, we add `candidate.coeff * candidate.expr = candidate.coeff * solution`
+            // to the constant.
+            constant += solution * candidate.coeff;
+            components[index] = Zero::zero();
+        }
+    }
+    if extracted_parts.is_empty() {
+        None
+    } else {
+        // We found some independent parts, add the remaining components to the parts
+        // and return them.
+        extracted_parts.push(recombine_rest(components, constant));
+        Some(extracted_parts)
+    }
+}
+
+/// Groups a sequence of components (thought of as a sum) by absolute coefficients
+/// so that its sum does not change.
+/// The list is sorted by the coefficient.
+fn group_components_by_absolute_coefficients<
+    T: RuntimeConstant + Display,
+    V: Clone + Ord + Display,
+>(
+    components: impl IntoIterator<Item = Component<T, V>>,
+) -> impl Iterator<Item = Component<T, V>> {
+    components
+        .into_iter()
+        .map(|c| c.normalize())
+        .into_grouping_map_by(|c| c.coeff)
+        .sum()
+        .into_iter()
+        .filter(|(_, expr)| !expr.is_zero())
+        .map(|(_, comp)| comp)
+        .sorted_by_key(|comp| comp.coeff.to_integer())
+}
+
+/// If this returns `Some(x)`, then `x` is the only valid value for `expr` in the equation
+/// `expr + rest + constant = 0`.
+/// It does not make assumptions about its inputs.
+fn find_solution<T: RuntimeConstant + Display, V: Clone + Ord + Display>(
+    expr: &GroupedExpression<T, V>,
+    rest: Vec<Component<T, V>>,
+    constant: T::FieldType,
+    range_constraints: &impl RangeConstraintProvider<T::FieldType, V>,
+) -> Option<T::FieldType> {
+    if rest.is_empty() {
+        // We are done anyway.
+        return None;
```

Comment:
isn't `-constant` the only valid value for `expr` in this case?

Reply:
Yes, we could return that, but it would be another code path to verify.

I could also use `?` in `let smallest_coeff = rest.iter().map(|comp| comp.coeff).min().unwrap();`...